File: src/net/merayen/elastic/ui/objects/components/Scroll.java
package net.merayen.elastic.ui.objects.components;

import net.merayen.elastic.ui.UIObject;
import net.merayen.elastic.ui.event.IEvent;
import net.merayen.elastic.ui.objects.UIClip;
import net.merayen.elastic.ui.util.Movable;

public class Scroll extends UIObject {
	private class Bar extends UIObject {
		private Movable movable;
		private boolean x;

		Bar(boolean x) {
			this.x = x;
		}

		@Override
		protected void onInit() {
			movable = new Movable(this, this);

			if(x) movable.drag_scale_y = 0;
			else movable.drag_scale_x = 0;

			movable.setHandler(new Movable.IMoveable() {
				@Override
				public void onMove() {
					updateFromBars();
				}

				@Override
				public void onGrab() {}

				@Override
				public void onDrop() {}
			});
		}

		@Override
		protected void onDraw() {
			draw.setColor(255, 0, 255);
			draw.fillRect(0, 0, bar_width, bar_width);
		}

		@Override
		protected void onEvent(IEvent e) {
			movable.handle(e);
		}
	}

	public float width = 100;
	public float height = 100;
	public float bar_width = 10;
	private final UIObject object;
	Scroll self = this;

	private final UIClip clip = new UIClip();
	private final Bar bar_x = new Bar(true);
	private final Bar bar_y = new Bar(false);

	private float content_width, content_height;

	public Scroll(UIObject object) {
		this.object = object;
	}

	@Override
	protected void onInit() {
		clip.add(object);
		add(clip);
		updateBars();
	}

	@Override
	protected void onDraw() {
		draw.setColor(0, 0, 0);
		draw.setStroke(bar_width);
		draw.fillRect(0, height - bar_width, width, bar_width);
		draw.fillRect(width - bar_width, 0, bar_width, height);

		draw.setColor(255, 0, 255);
	}

	@Override
	protected void onUpdate() {
		content_width = object.getWidth();
		content_height = object.getHeight();

		System.out.println(content_width + " " + content_height);

		clip.width = width - bar_width;
		clip.height = height - bar_width;

		if(object.translation.x > 0)
			object.translation.x = 0;

		if(object.translation.y > 0)
			object.translation.y = 0;

		if(object.translation.x < -(content_width - width) && content_width > width)
			object.translation.x = -(content_width - width);

		if(object.translation.y < -(content_height - height) && content_height > height)
			object.translation.y = -(content_height - height);

		bar_x.translation.y = height - bar_width;
		bar_y.translation.x = width - bar_width;

		updateBars();
	}

	private void updateBars() {
		if(content_width - width > 0) {
			if(bar_x.getParent() == null)
				add(bar_x);

			bar_x.translation.y = height - bar_width;
		} else if(bar_x.getParent() != null) {
			remove(bar_x);
		}

		if(content_height - height > 0) {
			if(bar_y.getParent() == null)
				add(bar_y);

			bar_y.translation.x = width - bar_width;
		} else if(bar_y.getParent() != null) {
			remove(bar_y);
		}
	}

	private void updateFromBars() {
		bar_x.translation.x = Math.max(0, Math.min(width - bar_width, bar_x.translation.x));
		bar_y.translation.y = Math.max(0, Math.min(height - bar_width, bar_y.translation.y));

		object.translation.x = ((bar_x.translation.x / (width - bar_width)) * -(content_width - width));
		object.translation.y = ((bar_y.translation.y / (height - bar_width)) * -(content_height - height));
	}
}

File: src/net/merayen/elastic/ui/objects/components/Scroll.kt
package net.merayen.elastic.ui.objects.components

import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.UIObject
import net.merayen.elastic.ui.event.UIEvent
import net.merayen.elastic.ui.objects.UIClip
import net.merayen.elastic.ui.util.Movable

class Scroll(private val uiobject: UIObject) : UIObject() {
    var layoutWidth = 100f
    var layoutHeight = 100f
    var barWidth = 10f

    private val clip = UIClip()
    private val barX = Bar(true)
    private val barY = Bar(false)

    private var contentWidth = 0f
    private var contentHeight = 0f
    private var moving = false

    private inner class Bar constructor(private val x: Boolean) : UIObject() {
        private val movable = Movable(this, this)

        override fun onInit() {
            if (x)
                movable.drag_scale_y = 0f
            else
                movable.drag_scale_x = 0f

            movable.setHandler(object : Movable.IMoveable {
                override fun onMove() {
                    updateFromBars()
                }

                override fun onGrab() {
                    moving = true
                }
                override fun onDrop() {
                    moving = false
                }
            })
        }

        override fun onDraw(draw: Draw) {
            draw.setColor(255, 0, 255)
            draw.fillRect(0f, 0f, barWidth, barWidth)
        }

        override fun onEvent(event: UIEvent) {
            movable.handle(event)
        }
    }

    override fun onInit() {
        clip.add(uiobject)
        add(clip)
        updateBars()
    }

    override fun onDraw(draw: Draw) {
        draw.setColor(50, 50, 50)
        draw.fillRect(0f, layoutHeight - barWidth, layoutWidth, barWidth)
        draw.fillRect(layoutWidth - barWidth, 0f, barWidth, layoutHeight)
    }

    override fun onUpdate() {
        contentWidth = uiobject.getWidth()
        contentHeight = uiobject.getHeight()

        clip.width = layoutWidth - barWidth
        clip.height = layoutHeight - barWidth

        if (uiobject.translation.x > 0f)
            uiobject.translation.x = 0f

        if (uiobject.translation.y > 0f)
            uiobject.translation.y = 0f

        if (uiobject.translation.x < -(contentWidth - layoutWidth) && contentWidth > layoutWidth)
            uiobject.translation.x = -(contentWidth - layoutWidth)

        if (uiobject.translation.y < -(contentHeight - layoutHeight) && contentHeight > layoutHeight)
            uiobject.translation.y = -(contentHeight - layoutHeight)

        barX.translation.y = layoutHeight - barWidth
        barY.translation.x = layoutWidth - barWidth

        if(!moving)
            updateBars()
    }

    private fun updateBars() {
        if (contentWidth - layoutWidth > 0) {
            if (barX.parent == null)
                add(barX)

            barX.translation.y = layoutHeight - barWidth
        } else if (barX.parent != null) {
            remove(barX)
        }

        if (contentHeight - layoutHeight > 0) {
            if (barY.parent == null)
                add(barY)

            barY.translation.x = layoutWidth - barWidth
        } else if (barY.parent != null) {
            remove(barY)
        }

        barX.translation.x = Math.max(0f, Math.min(layoutWidth - barWidth, barX.translation.x))
        barY.translation.y = Math.max(0f, Math.min(layoutHeight - barWidth, barY.translation.y))
    }

    private fun updateFromBars() {
        uiobject.translation.x = barX.translation.x / (layoutWidth - barWidth) * -(contentWidth - layoutWidth)
        uiobject.translation.y = barY.translation.y / (layoutHeight - barWidth) * -(contentHeight - layoutHeight)
    }
}
